dipy/denoise/denspeed.pyx

```diff
@@ -154,16 +158,19 @@ cdef double process_block(double [:, :, ::1] arr,
             for o in range(P, BS - P):
 
                 summ = 0
+                sigm = 0
 
                 # calculate square distance
-                for a in range(- P, P + 1):
-                    for b in range(- P, P + 1):
-                        for c in range(- P, P + 1):
+                for a in range(-P, P + 1):
+                    for b in range(-P, P + 1):
+                        for c in range(-P, P + 1):
 
                             # this line takes most of the time! mem access
                             d = cache[(B + a) * BS * BS + (B + b) * BS + (B + c)] - cache[(m + a) * BS * BS + (n + b) * BS + (o + c)]
                             summ += d * d
+                            sigm += sigma_block[(m + a) * BS * BS + (n + b) * BS + (o + c)]
```

Comment:
There are lot of computation going on here to obtain the index  `(m + a) * BS * BS` does not change when `b` and `c` do. The same goes for `(n + b) * BS` w.r.t. `c`. (See also line 169).

I don't know if the compiler already optimize this for you but if you need speed up it's worth trying.